### PR TITLE
Add metamask_watchAsset

### PR DESCRIPTION
**Pull request by danfinlay**
Adds the method `metamask_watchAsset`, a vendor-prefixed version of `wallet_watchAsset` per [EIP 747](https://ethereum-magicians.org/t/eip-747-wallet-watchasset/1048).

To Do:
- [x] Refine & finalize the EIP specification.
- [x] Fix rpc-engine crash preventing the confirmation from working correctly.
- [x] Add parameter for specifying token icon (maybe as a base64 string w/ a max length)

### Sample Usage

```javascript
web3.currentProvider.sendAsync({
  method: 'metamask_watchAsset',
  params: {
    "type":"ERC20",
    "options":{
      "address":"0x2e8c6bbe8e3aa934ee5a851b2cdff52412d61b86",
      "symbol":"FOX",
      "decimals":18,
      "image":'https://pbs.twimg.com/profile_images/802481220340908032/M_vde_oi_400x400.jpg'
    },
  },
  id: 20,
}, console.log)
```

### CODE REVIEW

**Comment by estebanmino** [comment]
Right now this PR is working as follow 

ezgif com-video-to-gif

@cjeria @danfinlay any thought on how it should be the cancellation of this suggestion?

**Comment by estebanmino** [comment]
ezgif com-video-to-gif 1

A new component was created to confirm the suggested tokens as you can see on the video. I think that could be a better approach for it. Thoughts?

Now we have the possibility to developers that want to use this feature to send a personal message through the method to the user, giving more information about this suggested token (?).

@danfinlay @cjeria

**Comment by danfinlay** [comment]
Okay, three thoughts:

- I'd like to get @cjeria to review the UX here. I think this might be worth some dedicated UI, since "back" and selecting other tokens is not really in the spirit of the feature.
- I'd like us to tuck the method under some kind of beta flag until we have an EIP and want to formalize: Maybe `metamask_watchToken`? Is that crazy?
- I'd like the method to support a token image also. Not sure whether it should be a bytes64 encoded jpeg, or a link to a hosted image or what. That might be a good question for an [ethereum-magicians](https://ethereum-magicians.org/) post. I'm happy to help draft that.

**Comment by estebanmino** [comment]
Following your comments @danfinlay 

- There are new components for this feature, in order to avoid the `Back` button there is a `Cancel` button that closes the notification. Working on new and old UI.

43747188-62926b74-99b6-11e8-894f-fbb2f4c79566
- Sounds good, not crazy for me at least.
- How do we continue with this?

File: app/scripts/controllers/preferences.js
Comment (brunobar79) [inline app/scripts/controllers/preferences.js]: Not sure if it's really needed but maybe use toChecksumAddress or normalizeAddress to make sure it will always find the right key (assuming that assetImages keys are already checksummed)

File: old-ui/app/add-suggested-token.js
Comment (brunobar79) [inline old-ui/app/add-suggested-token.js]: This should go on the messages.json file

File: old-ui/app/add-suggested-token.js
Comment (brunobar79) [inline old-ui/app/add-suggested-token.js]: and this one too

File: old-ui/app/add-suggested-token.js
Comment (brunobar79) [inline old-ui/app/add-suggested-token.js]: I'd say this one too but, are we translating "Token Symbol" this in other places?

File: old-ui/app/add-suggested-token.js
Comment (brunobar79) [inline old-ui/app/add-suggested-token.js]: This one for sure.

File: old-ui/app/add-suggested-token.js
Comment (brunobar79) [inline old-ui/app/add-suggested-token.js]: And this one

File: old-ui/app/add-suggested-token.js
Comment (brunobar79) [inline old-ui/app/add-suggested-token.js]: same

File: old-ui/app/add-suggested-token.js
Comment (brunobar79) [inline old-ui/app/add-suggested-token.js]: same

File: old-ui/app/add-suggested-token.js
Comment (brunobar79) [inline old-ui/app/add-suggested-token.js]: same

File: old-ui/app/add-suggested-token.js
Comment (brunobar79) [inline old-ui/app/add-suggested-token.js]: same

File: test/unit/app/controllers/preferences-controller-test.js
Comment (brunobar79) [inline test/unit/app/controllers/preferences-controller-test.js]: shouldn't ?

**Review by brunobar79 — COMMENTED** [review]
LGTM! just a few strings that should be moved to the localization files and a potential safe check.

File: app/scripts/controllers/preferences.js
Comment (bitpshr) [inline app/scripts/controllers/preferences.js]: Nit: You can save one level of indirection by doing `const { isValidAddress } = require('ethereumjs-util');`

**Review by bitpshr — DISMISSED** [review]
This looks great. There are a few strings that can be internationalized, but nothing show-stopping. Excited to see this feature in the wild.

One question: are we standardizing on the `metamask_` custom RPC method prefix? Should we version this as well in case new versions of the proposal arise?

File: old-ui/app/add-suggested-token.js
Comment (estebanmino) [reply old-ui/app/add-suggested-token.js]: I agree with all these comments about messages.json file, but the original old ui  (`old-ui/app/add-tokens.js`) all names are in this way and are not used to do translations, so I assumed that this wasn't necessary. Let me know what you think.

File: app/scripts/controllers/preferences.js
Comment (estebanmino) [reply app/scripts/controllers/preferences.js]: Yes, in line 295 the normalization happens, so for each token added it will have as key the correct normalized address.

**Comment by estebanmino** [comment]
Thanks for reviewing this.
@bitpshr the idea of that prefix is to have the first version of this before watchAsset is standardized on EM, we want to test and find room to improve it.
Waiting for @brunobar79 comments to merge.

File: old-ui/app/add-suggested-token.js
Comment (brunobar79) [reply old-ui/app/add-suggested-token.js]: If we don't move it now, we won't do it later, and that means that it would never get translated.

**Review by brunobar79 — DISMISSED** [review]
Looks good!  Ping me after resolving conflicts and I'll merge

**Review by brunobar79 — APPROVED** [review]

**Comment by not-idol** [comment]
Why is there a Character Limit for the symbol? e.g. Tokens with a character length of 8 can be imported into Metamask via the address but not with this method.

**Comment by df257** [comment]
I want add two token together, how to do it?
like this:
image

**Comment by PIKIINU** [comment]
web3.currentProvider.sendAsync({
  method: 'metamask_watchAsset',
  params: {
    "type":"ERC20",
    "options":{
      "address":"0x2e8c6bbe8e3aa934ee5a851b2cdff52412d61b86",
      "symbol":"FOX",
      "decimals":18,
      "image":'https://pbs.twimg.com/profile_images/802481220340908032/M_vde_oi_400x400.jpg'
    },
  },
  id: 20,
}, console.log)

**Comment by hiroshi-yamamoto-dublr** [comment]
Trying to add the same token twice gives the following error:

image

This warning should not be given if the token was already added to the wallet with the same `chainId` and token address (in fact the MetaMask window should probably not pop up at all in that case).

**Comment by WojasKrk** [comment]
> 
>   <details>
>     <summary>
>       Builds ready [56bed3f]:
>       <a href="https://49205-42009758-gh.circle-artifacts.com/0/builds/mascara/home.html">mascara</a>,
>       <a href="https://49205-42009758-gh.circle-artifacts.com/0/builds/metamask-chrome-4.9.3.zip">chrome</a>,
>       <a href="https://49205-42009758-gh.circle-artifacts.com/0/builds/metamask-firefox-4.9.3.zip">firefox</a>,
>       <a href="https://49205-42009758-gh.circle-artifacts.com/0/builds/metamask-edge-4.9.3.zip">edge</a>,
>       <a href="https://49205-42009758-gh.circle-artifacts.com/0/builds/metamask-opera-4.9.3.zip">opera</a>
>     </summary>
>     <image src="https://49205-42009758-gh.circle-artifacts.com/0/test-artifacts/screens/walkthrough%20%28en%29.gif">
>   </details>
>

File: test/unit/app/controllers/preferences-controller-test.js
Comment (WojasKrk) [reply test/unit/app/controllers/preferences-controller-test.js]: wojaskrk@gmail.com

**Pull request merged**
